### ISSUE TITLE
MPICH osu_get_bw fails with assertion in ch3_progress.c

### ISSUE THREAD

**Issue by sungeunchoi** [ofi-cray/libfabric-cray#726]
- srun -n2 -N2 -t00:2:00 --exclusive --cpu_bind=none osu_get_bw

Results in:

```
Assertion failed in file src/mpid/ch3/channels/nemesis/src/ch3_progress.c at line 758: pkt->type >= 0 && pkt->type < MPIDI_CH3_PKT_END_ALL
```

**Comment by hppritcha** [comment]
This test runs successfully, if slowly, using the sockets provider.

**Comment by hppritcha** [comment]
If I disable lazy memory deregistration and compile with --enable-debug, the signature changes.  I now get a failure in the gnix_mr code when an assert fails on the atomic cnt for a cache entry.  It appears somehow there's some corruption related to the check here in gnix_mr_cache.c

``` C
if (next != &entry->siblings && dlist_empty(next)) {
    parent = container_of(next, gnix_mr_cache_entry_t,
                                       children);
```

@jswaro could you explain the check?

**Comment by hppritcha** [comment]
I double checked that this behavior isn't impacted by whether or not auto or manual progress is used for control/data.  It appears some kind of memory corruption is happening in the mr cache when lazy_dregistration is enabled.  Even with lazy_deregistration disabled though, if one configures with --enable-debug set, one gets periodic assert failures for lock embedded in the `parent`.

**Comment by hppritcha** [comment]
I'm assigning this to @jswaro for further debugging.

**Comment by sungeunchoi** [comment]
Clarification for the comment above: the assert above might be a red herring due to not handling the actual error properly.

In any case, what is happening is that we are getting `GNI_RC_INVALID_PARAM` back from `GNI_MemRegister`.  I've confirmed that we are trying to register a memory region with the same starting address as one that was recently registered.  The first one is smaller than the second and was successfully registered.  User error seems unlikely tho I can't rule it out.  Note that the addresses I see have already been page aligned, so it probably it is the case that the new registration is a coalescing of adjacent memory regions where the smaller one is still in use.

**Comment by sungeunchoi** [comment]
I just saw the same error happen, but this time with a unique address that had never been registered before, though it does overlap with an existing registration.  All other parameters look reasonable (most had been used in previous successful registrations).  This points to the requested memory region containing unmapped memory.

This is either direct user error or some part of the memory region being freed by the user (and correctly not used, but still being coalesced).

**Comment by hppritcha** [comment]
I did also see similar errors before I started using --enable-debug.  I think that causes something to change that masks the `GNI_MemRegister` failures.

**Comment by sungeunchoi** [comment]
I am using --enable-debug.

**Comment by hppritcha** [comment]
move to 1.3.1 milestone

**Comment by sungeunchoi** [comment]
FWIW, running cleanly with the sockets provider doesn't prove anything, because memory registration is a no op.

**Comment by sungeunchoi** [comment]
Fixed with #809
